### ISSUE TITLE
Invoking location mode in a "nodaemon" Elvish binary panics

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1487]
I was experimenting with a "nodaemon" build of Elvish to explore whether using Asciinema is a better alternative to the current mechanism for generating a "ttyshot" and was surprised that pressing Alt-l to start location mode resulted in a panic. It doesn't matter whether or not a `cd` has been executed before pressing Alt-l:

```
> go build -o elvish-nodaemon ./cmd/nodaemon/elvish
> ./elvish-nodaemon
> cd /tmp
# Press Alt-l to invoke location mode
goroutine 1 [running]:
src.elv.sh/pkg/sys.DumpStack()
        /Users/krader/projects/3rd-party/elvish/pkg/sys/dumpstack.go:10 +0x66
src.elv.sh/pkg/shell.handlePanic()
        /Users/krader/projects/3rd-party/elvish/pkg/shell/interact.go:146 +0x5a
panic({0x42752a0, 0x44c0ea0})
        /usr/local/Cellar/go/1.17.5/libexec/src/runtime/panic.go:1038 +0x215
src.elv.sh/pkg/edit.dirStore.Dirs(...)
        /Users/krader/projects/3rd-party/elvish/pkg/edit/listing.go:259
src.elv.sh/pkg/cli/modes.NewLocation({0x431adc8, 0xc0000d8000}, {{0x430d120, 0xc0000a4810}, {0x43142
80, 0xc00000f5f0}, 0xc00000f5a8, 0xc00000f5d8, 0xc00009a528, {0x42ccb80, ...}})
        /Users/krader/projects/3rd-party/elvish/pkg/cli/modes/location.go:80 +0x214
src.elv.sh/pkg/edit.initLocation.func1()
        /Users/krader/projects/3rd-party/elvish/pkg/edit/listing.go:112 +0x259
reflect.Value.call({0x425be40, 0xc0000b0a80, 0x8e}, {0x42b7bdc, 0x4}, {0x0, 0x0, 0x406c4a0})
        /usr/local/Cellar/go/1.17.5/libexec/src/reflect/value.go:543 +0x814
reflect.Value.Call({0x425be40, 0xc0000b0a80, 0x4057e3f}, {0x0, 0x0, 0x0})
        /usr/local/Cellar/go/1.17.5/libexec/src/reflect/value.go:339 +0xc5
src.elv.sh/pkg/eval.(*goFn).Call(0xc0000b0af0, 0xc000250d00, {0x0, 0x0, 0x0}, 0xc000100a20)
        /Users/krader/projects/3rd-party/elvish/pkg/eval/go_fn.go:247 +0xb46
src.elv.sh/pkg/eval.(*Evaler).Call(0x400efcd, {0x430c980, 0xc0000b0af0}, {{0x0, 0x0, 0x0}, 0xc000100
a20, {0x42bb31e, 0x10}}, {{0xc00000f578, ...}, ...})
        /Users/krader/projects/3rd-party/elvish/pkg/eval/eval.go:434 +0x37f
src.elv.sh/pkg/edit.callWithNotifyPorts({0x43130e8, 0xc00008e060}, 0x1, {0x430c980, 0xc0000b0af0}, {
0x0, 0x0, 0x0})
        /Users/krader/projects/3rd-party/elvish/pkg/edit/key_binding.go:67 +0x1fb
src.elv.sh/pkg/edit.mapBindings.Handle({{0x43130e8, 0xc00008e060}, 0xc0000a8000, {0xc00009a1c8, 0x1,
 0x1}}, {0xc000677680, 0x400ff2e}, {0x430cfe0, 0xc0006e31b0})
        /Users/krader/projects/3rd-party/elvish/pkg/edit/key_binding.go:41 +0x117
src.elv.sh/pkg/cli/tk.(*codeArea).handleKeyEvent(0xc0000d8100, {0x677718, 0xc0})
        /Users/krader/projects/3rd-party/elvish/pkg/cli/tk/codearea.go:297 +0xd5
src.elv.sh/pkg/cli/tk.(*codeArea).Handle(0xc0000d8000, {0x430cfe0, 0xc0006e31a8})
        /Users/krader/projects/3rd-party/elvish/pkg/cli/tk/codearea.go:172 +0x72
src.elv.sh/pkg/cli.(*app).handle(0xc0000d8000, {0x4291d80, 0xc0006e31a8})
        /Users/krader/projects/3rd-party/elvish/pkg/cli/app.go:226 +0x165
src.elv.sh/pkg/cli.(*loop).Run(0xc0000cc100)
        /Users/krader/projects/3rd-party/elvish/pkg/cli/loop.go:129 +0x1b8
src.elv.sh/pkg/cli.(*app).ReadCode(0xc0000d8000)
        /Users/krader/projects/3rd-party/elvish/pkg/cli/app.go:475 +0x4a5
src.elv.sh/pkg/edit.(*Editor).ReadCode(0xc000010010)
        /Users/krader/projects/3rd-party/elvish/pkg/edit/editor.go:113 +0x22
src.elv.sh/pkg/shell.interact(0xc0000a8000, {0xc000010010, 0xc000010018, 0xc000010020}, 0xc000207d28
)
        /Users/krader/projects/3rd-party/elvish/pkg/shell/interact.go:103 +0x84a
src.elv.sh/pkg/shell.Program.Run({0x44fa7e0}, {0xc000010010, 0xc000010018, 0xc000010020}, 0xc0001671
f0, {0xc0000121e0, 0x2, 0x0})
        /Users/krader/projects/3rd-party/elvish/pkg/shell/shell.go:71 +0x2de
src.elv.sh/pkg/prog.compositeProgram.Run({0xc000100f90, 0x3, 0xc000111e50}, {0xc000010010, 0xc000010
018, 0xc000010020}, 0x410d16f, {0xc0000121e0, 0x0, 0x0})
        /Users/krader/projects/3rd-party/elvish/pkg/prog/prog.go:154 +0xd3
src.elv.sh/pkg/prog.Run({0xc000010010, 0xc000010018, 0xc000010020}, {0xc0000121e0, 0x1, 0x1}, {0x430
d8e0, 0xc00000e1f8})
        /Users/krader/projects/3rd-party/elvish/pkg/prog/prog.go:128 +0x406
main.main()
        /Users/krader/projects/3rd-party/elvish/cmd/nodaemon/elvish/main.go:15 +0xe6

goroutine 6 [chan receive]:
src.elv.sh/pkg/eval.getBlackholeChan.func1()
        /Users/krader/projects/3rd-party/elvish/pkg/eval/port.go:96 +0x3a
created by src.elv.sh/pkg/eval.getBlackholeChan
        /Users/krader/projects/3rd-party/elvish/pkg/eval/port.go:95 +0x6f

goroutine 7 [syscall]:
os/signal.signal_recv()
        /usr/local/Cellar/go/1.17.5/libexec/src/runtime/sigqueue.go:166 +0x28
os/signal.loop()
        /usr/local/Cellar/go/1.17.5/libexec/src/os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
        /usr/local/Cellar/go/1.17.5/libexec/src/os/signal/signal.go:151 +0x2c

goroutine 18 [chan receive]:
src.elv.sh/pkg/shell.initTTYAndSignal.func1()
        /Users/krader/projects/3rd-party/elvish/pkg/shell/shell.go:113 +0x86
created by src.elv.sh/pkg/shell.initTTYAndSignal
        /Users/krader/projects/3rd-party/elvish/pkg/shell/shell.go:112 +0x9f

goroutine 19 [chan receive]:
src.elv.sh/pkg/cli/prompt.(*Prompt).loop(0xc0000b0310)
        /Users/krader/projects/3rd-party/elvish/pkg/cli/prompt/prompt.go:77 +0xb6
created by src.elv.sh/pkg/cli/prompt.New
        /Users/krader/projects/3rd-party/elvish/pkg/cli/prompt/prompt.go:70 +0x1cf

goroutine 20 [chan receive]:
src.elv.sh/pkg/cli/prompt.(*Prompt).loop(0xc0000b03f0)
        /Users/krader/projects/3rd-party/elvish/pkg/cli/prompt/prompt.go:77 +0xb6
created by src.elv.sh/pkg/cli/prompt.New
        /Users/krader/projects/3rd-party/elvish/pkg/cli/prompt/prompt.go:70 +0x1cf

goroutine 7228 [semacquire]:
sync.runtime_Semacquire(0x44fa7e0)
        /usr/local/Cellar/go/1.17.5/libexec/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0x0)
        /usr/local/Cellar/go/1.17.5/libexec/src/sync/waitgroup.go:130 +0x71
src.elv.sh/pkg/eval.(*pipelineOp).exec.func2()
        /Users/krader/projects/3rd-party/elvish/pkg/eval/compile_effect.go:143 +0x5c
created by src.elv.sh/pkg/eval.(*pipelineOp).exec
        /Users/krader/projects/3rd-party/elvish/pkg/eval/compile_effect.go:142 +0x256

goroutine 7227 [syscall]:
syscall.syscall6(0x407ba40, 0x195a, 0xc0003fbb34, 0x0, 0xc000256090, 0x0, 0x0)
        /usr/local/Cellar/go/1.17.5/libexec/src/runtime/sys_darwin.go:44 +0x3b
syscall.wait4(0xc0003fbb38, 0x40126a7, 0x90, 0x42b0a80)
        /usr/local/Cellar/go/1.17.5/libexec/src/syscall/zsyscall_darwin_amd64.go:44 +0x48
syscall.Wait4(0xc0003fbb90, 0xc0003fbb6c, 0xc00083a280, 0x0)
        /usr/local/Cellar/go/1.17.5/libexec/src/syscall/syscall_bsd.go:145 +0x2b
os.(*Process).wait(0xc0000ca450)
        /usr/local/Cellar/go/1.17.5/libexec/src/os/exec_unix.go:44 +0x77
os.(*Process).Wait(...)
        /usr/local/Cellar/go/1.17.5/libexec/src/os/exec.go:132
src.elv.sh/pkg/eval.externalCmd.Call({{0xc00083a280, 0xf84}}, 0xc000780600, {0xc00063ed00, 0x5, 0xc0
000b7000}, 0xc000182420)
        /Users/krader/projects/3rd-party/elvish/pkg/eval/external_cmd.go:99 +0x2f9
src.elv.sh/pkg/eval.(*formOp).exec(0xc0004f6630, 0xc000780600)
        /Users/krader/projects/3rd-party/elvish/pkg/eval/compile_effect.go:366 +0x45b
src.elv.sh/pkg/eval.(*pipelineOp).exec.func1({0x430c960, 0xc0004f6630}, 0xc0006ca440)
        /Users/krader/projects/3rd-party/elvish/pkg/eval/compile_effect.go:120 +0x5a
created by src.elv.sh/pkg/eval.(*pipelineOp).exec
        /Users/krader/projects/3rd-party/elvish/pkg/eval/compile_effect.go:136 +0x749

runtime error: invalid memory address or nil pointer dereference
```

**Comment by krader1961** [comment]
Probably related to issue #1476. Note that a fix for that issue has been merged but because the commit message did not say it fixed that issue it is still open. Issue #1476 should probably be closed.

**Comment by krader1961** [comment]
As with issue #1476 the problem is the unqualified dereference of (dirStore).st here:

https://github.com/elves/elvish/blob/fda62502ad7f33222674a97602822114c4fd1037/pkg/edit/listing.go#L258-L260

The fix for the panic is obvious and trivial. However, it begs the question of why we can't have a location history when the daemon is not supported. Which prompts me to point to issue #1222. The Elvish daemon causes problems and only exists because of the use of the [Bbolt](https://pkg.go.dev/go.etcd.io/bbolt) database API which does not support concurrent access to its database from multiple processes.

**Comment by krader1961** [comment]
Note that I was building, and using, a "nodaemon" version of Elvish because I do not want someone's private history to leak into a "ttyshot" captured by Asciinema. Regardless of how this issue is resolved it should be possible to start an Elvish shell that does not inherit command/location history.

**Comment by krader1961** [comment]
FWIW, I have abandoned my use of a "nodaemon" Elvish binary when generating a "ttyshot". I have switched to explicitly starting an Elvish daemon with a pre-defined interactive database and telling the interactive Elvish process to talk to that daemon. Nonetheless, the fix for this issue should be merged since it is trivial and a panic is never acceptable and we don't want users of Elvish in the [`u-root`](https://github.com/u-root/u-root/) project to run into this failure.
